tcl/doc/Eval.3
'\"
'\" Copyright (c) 1989-1993 The Regents of the University of California.
'\" Copyright (c) 1994-1997 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH Tcl_Eval 3 8.1 Tcl "Tcl Library Procedures"
.BS
.SH NAME
Tcl_EvalObjEx, Tcl_EvalFile, Tcl_EvalObjv, Tcl_Eval, Tcl_EvalEx, Tcl_GlobalEval, Tcl_GlobalEvalObj, Tcl_VarEval, Tcl_VarEvalVA \- execute Tcl scripts
.SH SYNOPSIS
.nf
\fB#include <tcl.h>\fR
.sp
.VS
int
\fBTcl_EvalObjEx\fR(\fIinterp, objPtr, flags\fR)
.sp
int
\fBTcl_EvalFile\fR(\fIinterp, fileName\fR)
.sp
int
\fBTcl_EvalObjv\fR(\fIinterp, objc, objv, flags\fR)
.sp
int
\fBTcl_Eval\fR(\fIinterp, script\fR)
.sp
int
\fBTcl_EvalEx\fR(\fIinterp, script, numBytes, flags\fR)
.sp
int
\fBTcl_GlobalEval\fR(\fIinterp, script\fR)
.sp
int
\fBTcl_GlobalEvalObj\fR(\fIinterp, objPtr, flags\fR)
.sp
int
\fBTcl_VarEval\fR(\fIinterp, string, string, ... \fB(char *) NULL\fR)
.sp
int
\fBTcl_VarEvalVA\fR(\fIinterp, argList\fR)
.SH ARGUMENTS
.AS Tcl_Interp **termPtr;
.AP Tcl_Interp *interp in
Interpreter in which to execute the script.  The interpreter's result is
modified to hold the result or error message from the script.
.AP Tcl_Obj *objPtr in
A Tcl object containing the script to execute.
.AP int flags in
ORed combination of flag bits that specify additional options.
\fBTCL_EVAL_GLOBAL\fR and \fBTCL_EVAL_DIRECT\fR are currently supported.
.AP char *fileName in
Name of a file containing a Tcl script.
.AP int objc in
The number of objects in the array pointed to by \fIobjPtr\fR;
this is also the number of words in the command.
.AP Tcl_Obj **objv in
Points to an array of pointers to objects; each object holds the
value of a single word in the command to execute.
.AP int numBytes in
The number of bytes in \fIscript\fR, not including any
null terminating character.  If \-1, then all characters up to the
first null byte are used.
.AP char *script in
Points to first byte of script to execute.  This script must be in
writable memory: temporary modifications are made to it during
parsing.
.AP char *string in
String forming part of a Tcl script.
.AP va_list argList in
An argument list which must have been initialised using
\fBTCL_VARARGS_START\fR, and cleared using \fBva_end\fR.
.BE

.SH DESCRIPTION
.PP
The procedures described here are invoked to execute Tcl scripts in
various forms.
\fBTcl_EvalObjEx\fR is the core procedure and is used by many of the others.
It executes the commands in the script stored in \fIobjPtr\fR
until either an error occurs or the end of the script is reached.
If this is the first time \fIobjPtr\fR has been executed,
its commands are compiled into bytecode instructions
which are then executed.  The
bytecodes are saved in \fIobjPtr\fR so that the compilation step
can be skipped if the object is evaluated again in the future.
.PP
The return value from \fBTcl_EvalObjEx\fR (and all the other procedures
described here) is a Tcl completion code with
one of the values \fBTCL_OK\fR, \fBTCL_ERROR\fR, \fBTCL_RETURN\fR,
\fBTCL_BREAK\fR, or \fBTCL_CONTINUE\fR.
In addition, a result value or error message is left in \fIinterp\fR's
result; it can be retrieved using \fBTcl_GetObjResult\fR.
.PP
\fBTcl_EvalFile\fR reads the file given by \fIfileName\fR and evaluates
its contents as a Tcl script.  It returns the same information as
\fBTcl_EvalObjEx\fR.
If the file couldn't be read then a Tcl error is returned to describe
why the file couldn't be read.
.PP
\fBTcl_EvalObjv\fR executes a single pre-parsed command instead of a
script.  The \fIobjc\fR and \fIobjv\fR arguments contain the values
of the words for the Tcl command, one word in each object in
\fIobjv\fR.  \fBTcl_EvalObjv\fR evaluates the command and returns
a completion code and result just like \fBTcl_EvalObjEx\fR.
.PP
\fBTcl_Eval\fR is similar to \fBTcl_EvalObjEx\fR except that the script to
be executed is supplied as a string instead of an object and no compilation
occurs.  The string is parsed and executed directly (using
\fBTcl_EvalObjv\fR) instead of compiling it and executing the bytecodes.
In situations where it is known that the script will never be executed
again, \fBTcl_Eval\fR may be faster than \fBTcl_EvalObjEx\fR.
\fBTcl_Eval\fR returns a completion code and result just like
\fBTcl_EvalObjEx\fR.  Note: for backward compatibility with versions before
Tcl 8.0, \fBTcl_Eval\fR copies the object result in \fIinterp\fR to
\fIinterp->result\fR (use is deprecated) where it can be accessed directly.
This makes \fBTcl_Eval\fR somewhat slower than \fBTcl_EvalEx\fR, which
doesn't do the copy.
.PP
\fBTcl_EvalEx\fR is an extended version of \fBTcl_Eval\fR that takes
additional arguments \fInumBytes\fR and \fIflags\fR.  For the
efficiency reason given above, \fBTcl_EvalEx\fR is generally preferred
over \fBTcl_Eval\fR.
.PP
\fBTcl_GlobalEval\fR and \fBTcl_GlobalEvalObj\fR are older procedures
that are now deprecated.  They are similar to \fBTcl_EvalEx\fR and
\fBTcl_EvalObjEx\fR except that the script is evaluated in the global
namespace and its variable context consists of global variables only
(it ignores any Tcl procedures that are active).  These functions are
equivalent to using the \fBTCL_EVAL_GLOBAL\fR flag (see below).
.PP
\fBTcl_VarEval\fR takes any number of string arguments
of any length, concatenates them into a single string,
then calls \fBTcl_Eval\fR to execute that string as a Tcl command.
It returns the result of the command and also modifies
\fIinterp->result\fR in the same way as \fBTcl_Eval\fR.
The last argument to \fBTcl_VarEval\fR must be NULL to indicate the end
of arguments.  \fBTcl_VarEval\fR is now deprecated.
.PP
\fBTcl_VarEvalVA\fR is the same as \fBTcl_VarEval\fR except that
instead of taking a variable number of arguments it takes an argument
list. Like \fBTcl_VarEval\fR, \fBTcl_VarEvalVA\fR is deprecated.

.SH "FLAG BITS"
Any ORed combination of the following values may be used for the
\fIflags\fR argument to procedures such as \fBTcl_EvalObjEx\fR:
.TP 23
\fBTCL_EVAL_DIRECT\fR
This flag is only used by \fBTcl_EvalObjEx\fR; it is ignored by
other procedures.  If this flag bit is set, the script is not
compiled to bytecodes; instead it is executed directly
as is done by \fBTcl_EvalEx\fR.  The
\fBTCL_EVAL_DIRECT\fR flag is useful in situations where the
contents of an object are going to change immediately, so the
bytecodes won't be reused in a future execution.  In this case,
it's faster to execute the script directly.
.TP 23
\fBTCL_EVAL_GLOBAL\fR
If this flag is set, the script is processed at global level.  This
means that it is evaluated in the global namespace and its variable
context consists of global variables only (it ignores any Tcl
procedures at are active).

.SH "MISCELLANEOUS DETAILS"
.PP
During the processing of a Tcl command it is legal to make nested
calls to evaluate other commands (this is how procedures and
some control structures are implemented).
If a code other than \fBTCL_OK\fR is returned
from a nested \fBTcl_EvalObjEx\fR invocation,
then the caller should normally return immediately,
passing that same return code back to its caller,
and so on until the top-level application is reached.
A few commands, like \fBfor\fR, will check for certain
return codes, like \fBTCL_BREAK\fR and \fBTCL_CONTINUE\fR, and process them
specially without returning.
.PP
\fBTcl_EvalObjEx\fR keeps track of how many nested \fBTcl_EvalObjEx\fR
invocations are in progress for \fIinterp\fR.
If a code of \fBTCL_RETURN\fR, \fBTCL_BREAK\fR, or \fBTCL_CONTINUE\fR is
about to be returned from the topmost \fBTcl_EvalObjEx\fR
invocation for \fIinterp\fR,
it converts the return code to \fBTCL_ERROR\fR
and sets \fIinterp\fR's result to an error message indicating that
the \fBreturn\fR, \fBbreak\fR, or \fBcontinue\fR command was
invoked in an inappropriate place.
This means that top-level applications should never see a return code
from \fBTcl_EvalObjEx\fR other then \fBTCL_OK\fR or \fBTCL_ERROR\fR.
.VE

.SH KEYWORDS
execute, file, global, object, result, script



CHANGELOG
Release NG4.2 (Mon Mar 29 19:37:20 CEST 2010)

(INTERNAL) Add Ray Minnerly to CONTRIBUTORS
(BUGFIX) Fix "can't save modified files" bug
(BUGFIX) Fix non-portable globbing via $path_sep, use "file join" instead


Release NG4.1 (Tue Feb  2 21:07:20 CET 2010)

(INTERNAL) Reorganize CONTRIBUTORS, add Maurice Havelday
(INTERNAL) Merge 672:747 of https://svn.berlios.de/svnroot/repos/sourcenav/branches/SN-NG4-phpv5_support for basic php5 support
(BUGFIX) Try to be more careful when glob'ing files and paths in sn_convert_FileName
(BUGFIX) Fix "no event type or button # or keysym" tcl/tk error on Fedora
(INTERNAL) Add John Mills to CONTRIBUTORS
(BUGFIX) Fix rendering of project selection dialog on FC10
(BUGFIX) Remove unneeded cscope file (db4/dist/cscope.out)
(INTERNAL) Properly AM_INIT_AUTOMAKE with package name snavigator instead of sn


Release NG4 (Thu Feb  5 13:28:54 CET 2009)

(INTERNAL) Reformat snptools.c to sensible K&R style
(INTERNAL) Improve parser debugging by adding generic help output to the parser skeleton snavigator/parsers/toolbox/snptools.c
(BUGFIX) Do not display "cancel" in external editor is not defined" error popup, but only OK
(ENHANCE) Tidy up code flow in sn_open_project() for further actions
(ENHANCE) Add forced project unlocking by user, part I
(ENHANCE) Add forced project unlocking by user, part II
(ENHANCE) Add unlocking of project files via command-line
(ENHANCE) Make sn_log() for the loading of language file more explicit
(BUGFIX) Space-tidy-up sn_prop.cfg.in for improved grouping and readability before major work
(BUGFIX) Against list --with-mingw in the top-level configure script
(BUGFIX) Add scrollbars for "files not found - what do you want to do" project refresh dialogue
(BUGFIX) Make startup semantics of snavigator.exe similar to unix and wait for hyper to exit
(INTERNAL) Remove links in db4/PORT and copy real files there to unbreak win32 building, part I
(INTERNAL) Remove links in db4/PORT and copy real files there to unbreak win32 building, part II
(INTERNAL) Remove links in db4/PORT and copy real files there to unbreak win32 building, part III
(INTERNAL) Merge https://svn.berlios.de/svnroot/repos/sourcenav/branches/SN-NG3-php_class_support, r634:666
(INTERNAL) Up version to NG4
(BUGFIX) Make snpdbg (parser debugging startup) work again
(BUGFIX) Rename button "Cancel" in project selection dialogue to "Quit"
(INTERNAL) Merge https://svn.berlios.de/svnroot/repos/sourcenav/branches/SN-NG3-new_gui_colors, r649:659
(BUGFIX) Remove passus about the SN logo from copyright notice - it's ours
(BUGFIX) Remove colourful borders around widgets (highlight-bg)
(ENHANCE) Re-add chill parser .[ch] files to repo, generated by yacc
(ENHANCE) Enhance (x)emacs functionality with SN
(BUGFIX) Fix building under win32/MinGW/msys for tix/ subdir
(ENHANCE) Merge SN-NG3-fsstnd branch and improve fsstnd compliance, part I
(BUGFIX) Remove reference to parsers of "Source-Navigator 4.0" in demo/asm/README
(INTERNAL) Change official sourcenav.ico to newer logo for NG series
(BUGFIX) fix dbimp startup on win32 - typo with sn_path(parserdirdir) instead of sn_path(parserdir)
(INTERNAL) Change official sourcenav.ico to newer logo for NG series, part II
(INTERNAL) Change official sourcenav.ico to newer logo for NG series, part III
(INTERNAL) Update copyright for SN NG to include the years 2007-2009
(BUGFIX) Fix win32 startup
(INTERNAL) Use AC_MSG_WARN instead of AC_MSG_ERROR when checking for a program with a specific version
(INTERNAL) Blacklist flex 2.5.33 which is known to break some parsers
(BUGFIX) Properly indent and format snavigator/parsers/common/lexinput.c before changing
(BUGFIX) Fix parser termination on lone LF encounter
(INTERNAL) Remove debug output for paths initialization and parser addition
(BUGFIX) Manually fix run-away ticks for configure and some m4s again (tcl,tk)
(INTERNAL) Make INSTALL notes more clear for win32 building w/ msys/mingw
(INTERNAL) Fix README wrongly referencing NEWS instead of CHANGELOG for changes


Release NG3 (Sat Feb  2 21:59:12 CET 2008)

(BUGFIX) Fix tcl rng() problems on 64bit platforms
(INTERNAL) Add Daniel to CONTRIBUTORS
(INTERNAL) delete +x from snavigator/hyper/tclfileutils.c
(INTERNAL) Remove +x from other .c files
(INTERNAL) Remove +x from .h .xbm .gif .html files
(INTERNAL) Remove +x from other non-executables
(INTERNAL) Update TODO, remove already achieved goal
(ENHANCE) Pretty up snavigator --help output
(ENHANCE) Activate "Open in external editor" feature for editorpane, part I
(ENHANCE) Activate "Open in external editor" feature for editorpane, part II
(ENHANCE) Activate "Open in external editor" feature for editorpane, part III
(ENHANCE) Activate "Open in external editor" feature for editorpane, part III
(ENHANCE) Activate "Open in external editor" feature for editorpane, part IV
(INTERNAL) Add a paragraph about source layout in INSTALL
(INTERNAL) Move snavigator/README.TXT to ./README
(INTERNAL) Alter README to account for SN NG and its homepage
(INTERNAL) Move snavigator/README.TXT to ./README, part II
(INTERNAL) Remove installing of snavigator/NEWS
(ENHANCE) Migrate from db-1.85 to db-4.6.21
(INTERNAL) Up version to NG3
(INTERNAL) Adjust db/ to db4/ in INSTALL in source layout paragraph
(ENHANCE) Tell SN to use the phpbrowser for .php3 and .php4 files also
(BUGFIX) Make distclean more clean for tix/
(BUGFIX) Fix INSTALL to clearly point out that out-of-tree building is currently not working


Release NG2 (Tue Sep 11 14:07:25 CEST 2007)

(INTERNAL) cleanup (K&R'ify) dbimp.c source
(BUGFIX) Fix building in seperate build directory for bitmaps
(BUGFIX) Remove -fwritable-strings from CFLAGS which caused build problems on gcc 4+
(INTERNAL) Move INSTALL.TXT into ./INSTALL and change default doc for building to in-source-tree.
(INTERNAL) Bugfix move of snavigator/INSTALL.TXT to ./INSTALL
(BUGFIX) Fix tcl + tk configure apostrophe runaways that prevented building
(ENHANCE) New artwork
(ENHANCE) New artwork II
(INTERNAL) Up version to NG2

snavigator/parsers/verilog/eval.cc
/*
 * Copyright (c) 1998-1999 Stephen Williams (steve@icarus.com)
 *
 *    This source code is free software; you can redistribute it
 *    and/or modify it in source code form under the terms of the GNU
 *    General Public License as published by the Free Software
 *    Foundation; either version 2 of the License, or (at your option)
 *    any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */
#if !defined(WINNT) && !defined(macintosh)
#ident "$Id: eval.cc,v 1.1 2000/12/21 21:57:13 jrandrews Exp $"
#endif

# include  "PExpr.h"
# include  "netlist.h"

verinum* PExpr::eval_const(const Design*, const string&) const
{
      return 0;
}

verinum* PEBinary::eval_const(const Design*des, const string&path) const
{
      verinum*l = left_->eval_const(des, path);
      if (l == 0) return 0;
      verinum*r = right_->eval_const(des, path);
      if (r == 0) {
	    delete l;
	    return 0;
      }

      verinum*res;

      switch (op_) {
	  case '+': {
		assert(l->is_defined());
		assert(r->is_defined());
		long lv = l->as_long();
		long rv = r->as_long();
		res = new verinum(lv+rv, l->len());
		break;
	  }
	  case '-': {
		assert(l->is_defined());
		assert(r->is_defined());
		long lv = l->as_long();
		long rv = r->as_long();
		res = new verinum(lv-rv, l->len());
		break;
	  }
	  case '*': {
		assert(l->is_defined());
		assert(r->is_defined());
		long lv = l->as_long();
		long rv = r->as_long();
		res = new verinum(lv * rv, l->len());
		break;
	  }
	  case '/': {
		assert(l->is_defined());
		assert(r->is_defined());
		long lv = l->as_long();
		long rv = r->as_long();
		res = new verinum(lv / rv, l->len());
		break;
	  }
	  case '%': {
		assert(l->is_defined());
		assert(r->is_defined());
		long lv = l->as_long();
		long rv = r->as_long();
		res = new verinum(lv % rv, l->len());
		break;
	  }
	  default:
	    delete l;
	    delete r;
	    return 0;
      }

      return res;
}

/*
 * Evaluate an identifier as a constant expression. This is only
 * possible if the identifier is that of a parameter.
 */
verinum* PEIdent::eval_const(const Design*des, const string&path) const
{
      const NetScope*scope = des->find_scope(path);
      assert(scope);
      const NetExpr*expr = des->find_parameter(scope, text_);

      if (expr == 0)
	    return 0;

      assert(msb_ == 0);

      if (dynamic_cast<const NetEParam*>(expr)) {
	    cerr << get_line() << ": sorry: I cannot evaluate ``" <<
		  text_ << "'' in this context." << endl;
	    return 0;
      }

      const NetEConst*eval = dynamic_cast<const NetEConst*>(expr);
      assert(eval);
      return new verinum(eval->value());
}

verinum* PENumber::eval_const(const Design*, const string&) const
{
      return new verinum(value());
}

verinum* PETernary::eval_const(const Design*des, const string&path) const
{
      verinum*test = expr_->eval_const(des, path);
      if (test == 0)
	    return 0;

      verinum::V bit = test->get(0);
      delete test;
      switch (bit) {
	  case verinum::V0:
	    return fal_->eval_const(des, path);
	  case verinum::V1:
	    return tru_->eval_const(des, path);
	  default:
	    return 0;
	      // XXXX It is possible to handle this case if both fal_
	      // and tru_ are constant. Someday...
      }
}


/*
 * $Log: eval.cc,v $
 * Revision 1.1  2000/12/21 21:57:13  jrandrews
 * initial import
 *
 * Revision 1.14  2000/03/08 04:36:53  steve
 *  Redesign the implementation of scopes and parameters.
 *  I now generate the scopes and notice the parameters
 *  in a separate pass over the pform. Once the scopes
 *  are generated, I can process overrides and evalutate
 *  paremeters before elaboration begins.
 *
 * Revision 1.13  2000/02/23 02:56:54  steve
 *  Macintosh compilers do not support ident.
 *
 * Revision 1.12  1999/11/30 04:48:17  steve
 *  Handle evaluation of ternary during elaboration.
 *
 * Revision 1.11  1999/11/28 23:42:02  steve
 *  NetESignal object no longer need to be NetNode
 *  objects. Let them keep a pointer to NetNet objects.
 *
 * Revision 1.10  1999/11/21 20:03:24  steve
 *  Handle multiply in constant expressions.
 *
 * Revision 1.9  1999/10/08 17:48:09  steve
 *  Support + in constant expressions.
 *
 * Revision 1.8  1999/09/20 02:21:10  steve
 *  Elaborate parameters in phases.
 *
 * Revision 1.7  1999/09/18 01:52:48  steve
 *  Remove spurious message.
 *
 * Revision 1.6  1999/09/16 04:18:15  steve
 *  elaborate concatenation repeats.
 *
 * Revision 1.5  1999/08/06 04:05:28  steve
 *  Handle scope of parameters.
 *
 * Revision 1.4  1999/07/17 19:51:00  steve
 *  netlist support for ternary operator.
 *
 * Revision 1.3  1999/05/30 01:11:46  steve
 *  Exressions are trees that can duplicate, and not DAGS.
 *
 * Revision 1.2  1999/05/10 00:16:58  steve
 *  Parse and elaborate the concatenate operator
 *  in structural contexts, Replace vector<PExpr*>
 *  and list<PExpr*> with svector<PExpr*>, evaluate
 *  constant expressions with parameters, handle
 *  memories as lvalues.
 *
 *  Parse task declarations, integer types.
 *
 * Revision 1.1  1998/11/03 23:28:58  steve
 *  Introduce verilog to CVS.
 *
 */

